include/deal.II/base/tensor_base.h

```diff
@@ -116,14 +115,28 @@ class Tensor<0,dim,Number>
   typedef typename numbers::NumberTraits<Number>::real_type real_type;
 
   /**
-   * Constructor. Set to zero.
+   * The tensor type this object represents. In the special case of a
+   * tensor or rank 0 we strip the tensor class and just store the scalar
+   * object.
    */
-  Tensor ();
+  typedef Number tensor_type;
```

Comment:
```
typedef Number tensor_type;
```
Well, this is just a mechanism to preserve the old logic - namely saving the underlying data in ```Tensor<1,dim,Number>``` as a plain array.

My suggestion for a further cleanup would be to remove this mechanism altogether and always store a tensor type in ```values```. This would also remove the necessity for half of the ```static_cast<>```s.

I guess we can trust a modern compiler enough to store an array of ```Tensor<0,dim,Number>``` efficiently.